### ISSUE TITLE
Bluebird warning about a non returned promise while using queryBuilder.page()

### ISSUE THREAD

**Issue by lostdalek** [Vincit/objection.js#295]
**EDIT: I discovered it has been fixed in v0.7.x, is it possible to release a v0.6.x with this fix?**

Hi!
I've got the following warning when I'm creating a paginated query:

```
(node:7774) Warning: a promise was created in a handler at eval at createHookCaller (/Users/florian/workspace/hapi-api/node_modules/objection/lib/queryBuilder/QueryBuilder.js:1429:16), <anonymous>:6:19 but was not returned from it, see http://goo.gl/rRqMUw
	at Function.Promise.attempt.Promise.try (/Users/florian/workspace/hapi-api/node_modules/bluebird/js/release/method.js:29:9)
```

the warning seems to comes from `line 494` in `QueryBuilder.js` (`Range` method), adding a "`return null;`" make the warning disappear.

I can make a PR but I'm not sure of the implication of this change (sounds like it shouldn't be a problem...)

objectionJs: v0.6.2

**Comment by jeff-kilbride** [comment]
I am also seeing these warnings. As a temporary fix, I was able to turn them off by setting:

BLUEBIRD_W_FORGOTTEN_RETURN=0

in my environment variables. However, I am also getting warnings whenever I rollback a transaction:

```
(node:18446) Warning: a promise was rejected with a non-error: [object Undefined]
```

Unfortunately, the error isn't very informative.

**Comment by devinivy** [comment]
If you run the process with the node flag `--trace-warnings`, might that help track it down?

**Comment by jeff-kilbride** [comment]
Yep, that did it! Learn something new every day...

```
(node:18460) Warning: a promise was rejected with a non-error: [object Undefined]
    at [...]/node_modules/knex/lib/dialects/mysql/transaction.js:50:27
From previous event:
    at [...]/node_modules/knex/lib/transaction.js:93:14
    at runCallback (timers.js:649:20)
    at tryOnImmediate (timers.js:622:5)
    at processImmediate [as _immediateCallback] (timers.js:594:5)
From previous event:
    at Transaction_MySQL.Transaction ([...]/node_modules/knex/lib/transaction.js:64:41)
    at new Transaction_MySQL ([...]/node_modules/knex/lib/dialects/mysql/transaction.js:32:25)
    at Client_MySQL.transaction ([...]/node_modules/knex/lib/dialects/mysql/index.js:88:12)
    at Function.transaction ([...]/node_modules/knex/lib/util/make-knex.js:75:21)
    at [...]/node_modules/objection/lib/transaction.js:117:10
From previous event:
    at Function.transaction.start ([...]/node_modules/objection/lib/transaction.js:116:10)
    at Object.<anonymous> ([...]/src/api/controllers/ApSiteController.js:44:47)
    at next (native)

(edited for brevity)
```

It's pointing to this line in my code:

```
trx = yield objection.transaction.start(Site.knex());
```

**Comment by fl0w** [comment]
Since you're yielding, I'm assuming this is wrapped in a coroutine?

**Comment by jeff-kilbride** [comment]
Yes. Bluebird's `coroutine`. And `trx` is defined prior to this line. This appears in a `try...catch`, so `trx` is defined outside the `try` in order to be available in the `catch`.

**Comment by devinivy** [comment]
Looks like it's almost definitely coming from knex.

**Comment by jeff-kilbride** [comment]
```
├── knex@0.12.6
```

**Comment by fl0w** [comment]
@jeff-kilbride just my sanity, could you try the same thing with knex `0.12.5`? The the QueryBuilder returns null shouldn't be an issue as it's wrapped with a Bluebird.try.

**Comment by devinivy** [comment]
Here it is– rejecting with a non-error value– https://github.com/tgriesser/knex/blob/master/src/dialects/mysql/transaction.js#L34

Is your transaction failing?

**Comment by jeff-kilbride** [comment]
Yeah, it's during `mocha` testing. I'm deliberately causing validation errors and, in some parts of my code, using `sinon-as-promised` to stub the model method and reject the promise.

**Comment by jeff-kilbride** [comment]
@fl0w  Sorry, just saw your post. Do you need me to test with `0.12.5`?

**Comment by jeff-kilbride** [comment]
For example:

```
        it('should test the try...catch with an error', function () {

            // Stub the Site.getByAccountId method.
            const gbai = sinon.stub(Site, 'getByAccountId');
            gbai.rejects(Utils.CError.get('foobar'));

            return request(sails.hooks.http.app)
                .get('/ap/site/list')
                .set('Accept', 'application/json')
                .set('signature', Utils.getSignature(query, authData.nonce.pri))
                .query(query)
                .expect(400, 'ERROR: foobar.')
                .expect(function () {
                    gbai.restore();
                });
        });
```

**Comment by jeff-kilbride** [comment]
This is `QueryBuilder.js:1424-1437` in the transpiled `objection` code:
```js
function createHookCaller(hook) {
  var hasMethod = 'has' + _lodash2.default.upperFirst(hook);

  // Compile the caller function for (measured) performance boost.
  var caller = new Function('promise', 'op', '\n    if (op.' + hasMethod + '()) {\n      return promise.then(function (result) {\n        return op.' + hook + '(this.builder, result);\n      });\n    } else {\n      return promise;\n    }\n  ');

  return function (promise, operations) {
    for (var i = 0, l = operations.length; i < l; ++i) {
      promise = caller(promise, operations[i]);
    }

    return promise;
  };
}
```
Somebody from the `objection` team will need to provide context for it, though...

**Comment by fl0w** [comment]
@jeff-kilbride I apologise, didn't mean for you to src hunting, I can find that myself. I meant whatever was invoking that in the first place.

Did you try with knex 0.12.5?

**Comment by jeff-kilbride** [comment]
@fl0w Not yet, and probably won't be able to until this evening. I'm on my way out, right now. I will post results here when I get a chance.

**Comment by jeff-kilbride** [comment]
@fl0w  I was able to squeeze this in...

```
├── knex@0.12.5
```

Looks identical, so I hope I downgraded correctly:

```
(node:18728) Warning: a promise was rejected with a non-error: [object Undefined]
    at [...]/node_modules/knex/lib/dialects/mysql/transaction.js:50:27
From previous event:
    at [...]/node_modules/knex/lib/transaction.js:93:14
    at runCallback (timers.js:649:20)
    at tryOnImmediate (timers.js:622:5)
    at processImmediate [as _immediateCallback] (timers.js:594:5)
From previous event:
    at Transaction_MySQL.Transaction ([...]/node_modules/knex/lib/transaction.js:64:41)
    at new Transaction_MySQL ([...]/node_modules/knex/lib/dialects/mysql/transaction.js:32:25)
    at Client_MySQL.transaction ([...]/node_modules/knex/lib/dialects/mysql/index.js:88:12)
    at Function.transaction ([...]/node_modules/knex/lib/util/make-knex.js:75:21)
    at [...]/node_modules/objection/lib/transaction.js:117:10
From previous event:
    at Function.transaction.start ([...]/node_modules/objection/lib/transaction.js:116:10)
    at Object.<anonymous> ([...]/src/api/controllers/ApSiteController.js:44:47)
    at next (native)
```

**Comment by fl0w** [comment]
It's hard for me to guess what's going on here without a minimal reproducible example. I haven't been using objection.js for a while (in-house politics), thus I'm not well read on changes between 0.6.x and 0.7 to know if there's anything in lib that might invoke this.

**Comment by koskimas** [comment]
The original issue has been fixed in 0.7.0 and I'm not going to fix it for earlier versions since this is not critical. Please open separate issues for the other cases if needed.

**Comment by jeff-kilbride** [comment]
Just for posterity...

I think I finally figured this out. While stepping through some unrelated stuff with the debugger, I noticed that the signature for the transaction rollback function was:
```javascript
function rollback(error) { ...
```
In the code above, generating the warnings, I was calling rollback like so:
```javascript
let trx;
try {
    trx = yield objection.transaction.start(Model.knex());
    // Do several things with the transaction object.
    yield trx.commit();
}
catch(e) {
    yield trx.rollback();
    // Log the error...
}
```
Changing the call to rollback to the following:
```javascript
yield trx.rollback(e);
```
got rid of all the `a promise was rejected with a non-error` warnings. It's not very clear anywhere that rollback takes an error object as a parameter...
